### ISSUE TITLE
Incorrect derivative from bug in jitrules.jl: nonzero `dval` added twice

### ISSUE THREAD

**Issue by danielwe** [EnzymeAD/Enzyme.jl#2182]
Certain patterns of splatting/tuple concatenation along the lines of `test/mixedapplyiter.jl` lead to incorrect derivatives when a Duplicated's `dval` is nonzero. Instead of `dval .+= dret`, you effectively get `dval .+= dval .+ dret`. The relevant tests always use `dval .= 0`, concealing the bug.

MWE:

```julia
using Enzyme

concat() = ()
concat(a) = a
concat(a, b) = (a..., b...)
concat(a, b, c...) = concat(concat(a, b), c...)

function f(x)
    t = concat(x...)  # fails
    # t = x[1]        # works
    return t[1] * t[1] + t[2][1] * t[2][1]
end

x = [(2.0, [2.7])]
dx = [(1.0, [1.0])]
res = Enzyme.autodiff(Reverse, f, Active, Duplicated(x, dx))
dx_true = [((5.0, [6.4]),)]
@show dx
@show dx_true
```
```julia-repl
julia> include("bug.jl");
dx = [(6.0, [6.4])]
dx_true = [((5.0, [6.4]),)]
```

The double accumulation happens in the following function:

https://github.com/EnzymeAD/Enzyme.jl/blob/3b36ea25157efd37ebe63d1e025decb6defaeb43/src/rules/jitrules.jl#L1395-L1404

When the MWE above reaches this call, `vec == dx`, `val == dx[1]`, and `expr == dx_true`. That is, `expr` already contains the correct final value of `dx` including the contribution from the initial values in `dx`, so applying `recursive_add` to add them up again is incorrect. (The array-wrapped value is not affected by the bug because `guaranteed_nonactive` is passed to `recursive_add`, stopping the recursion at `DupState` values.)

I've verified that the MWE is fixed if I change line 1397 to
```julia
@inbounds vec[idx_in_vec] = expr
```
The entire test suite also passes after this change. However, while I haven't fully wrapped my head around `jitrules.jl`, it doesn't look like this would be the proper solution.

**Comment by danielwe** [comment]
Simpler MWE:

```julia
using Enzyme

concat() = ()
concat(a) = a
concat(a, b) = (a..., b...)
concat(a, b, c...) = concat(concat(a, b), c...)

f(x) = first(concat(x...))

x = [(2.0, [2.7])]
dx = [(1.0, [1.0])]
res = Enzyme.autodiff(Reverse, f, Active, Duplicated(x, dx))
dx_true = [(2.0, [1.0])]
@show dx
@show dx_true
```

The problem appears in the runtime handling of the type unstable function `concat` in the reverse pass. Can confirm that just setting `vec[idx_in_vec] = expr` is not the solution; this gives incorrect answers in cases where the arguments to `concat` are `Active` instead of `MixedDuplicated` (apparently no tests hit this functionality at the moment).

I wonder if the error is that the following code handles `MixedDuplicated` at all?

https://github.com/EnzymeAD/Enzyme.jl/blob/3b36ea25157efd37ebe63d1e025decb6defaeb43/src/rules/jitrules.jl#L1454-L1479

I get that the idea is to accumulate shadows for active args, including active parts of MixedDuplicated, but at least in the MWE it seems like MixedDuplicated shadows have already been handled in-place by the time we get here, nothing more to do. Is the solution simply to remove the (Batch)MixedDuplicated branches from this part?

**Comment by wsmoses** [comment]
Yeah you're absolutely right I don't think we should update anything for the mixedduplicated here. Make a PR (and definitely add the test)?
